mlflow/store/dynamodb_store.py

```diff
@@ -0,0 +1,707 @@
+import os
+import uuid
+
+import boto3
+from boto3.dynamodb.conditions import Key, And
+from decimal import Decimal
+
+from mlflow.entities import Experiment, Metric, Param, Run, RunData, RunInfo, RunStatus, RunTag, \
+                            ViewType
+from mlflow.entities.run_info import check_run_is_active, \
+    check_run_is_deleted
+from mlflow.exceptions import MlflowException
+import mlflow.protos.databricks_pb2 as databricks_pb2
+from mlflow.store.abstract_store import AbstractStore
+from mlflow.utils.validation import _validate_metric_name, _validate_param_name, _validate_run_id, \
+                                    _validate_tag_name
+
+from mlflow.utils.env import get_env
+from mlflow.utils.mlflow_tags import MLFLOW_RUN_NAME, MLFLOW_PARENT_RUN_ID
+
+from mlflow.utils.search_utils import does_run_match_clause
+
+_DYNAMODB_ENDPOINT_URL_VAR = "MLFLOW_DYNAMODB_ENDPOINT_URL"
+_DYNAMODB_TABLE_PREFIX_VAR = "MLFLOW_DYNAMODB_TABLE_PREFIX"
+
+
+def _default_dynamodb_resource(endpoint_url=None):
+    dynamodb_endpoint_url = endpoint_url or get_env(_DYNAMODB_ENDPOINT_URL_VAR)
+    return boto3.resource('dynamodb', endpoint_url=dynamodb_endpoint_url)
+
+
+def _default_table_prefix():
+    return get_env(_DYNAMODB_TABLE_PREFIX_VAR) or "mlflow"
+
+
+def _dict_to_experiment(d):
+    return Experiment(
+        experiment_id=int(d['experiment_id']),
+        name=d['name'],
+        artifact_location=d.get('artifact_location') or '',
+        lifecycle_stage=d.get('lifecycle_stage', 'active')
+    )
+
+
+def _dict_to_run_info(d):
+    return RunInfo(
+        run_uuid=d['run_uuid'],
+        experiment_id=int(d['experiment_id']),
+        name=d.get('name') or '',
+        source_type=int(d.get('source_type') or 0),
+        source_name=d.get('source_name') or '',
+        entry_point_name=d.get('entry_point_name') or '',
+        user_id=d.get('user_id') or '',
+        status=int(d.get('status') or 0),
+        start_time=int(d.get('start_time') or 0),
+        end_time=int(d.get('end_time') or 0),
+        source_version=d.get('source_version') or '',
+        lifecycle_stage=d.get('lifecycle_stage', 'active'),
+        artifact_uri=d.get('artifact_uri') or '',
+    )
+
+
+def _list_to_run_tag(l):
+    return [RunTag(
+        key=d['key'],
+        value=d['value']
+    ) for d in l]
+
+
+def _list_to_run_param(l):
+    return [Param(
+        key=d['key'],
+        value=d['value']
+    ) for d in l]
+
+
+def _sort_metrics(l):
+    return sorted(l, key=lambda x: int(x['timestamp']))
+
+
+# Return a list of metrics with the first value (don't sort)
+def _list_to_run_metric(l):
+    return [Metric(key=rm['key'],
+            value=float(rm['metrics'][0]['value']),
+            timestamp=int(rm['metrics'][0]['timestamp'])) for rm in l]
```

Comment:
What metric value is the first element? It should be the last recorded metric. 

(I see you apply sort when looking for metric history. I am confused why do you need it there and not here)